File: src/2025-02-24-deepclaude-rust/go/exercises/practice/bottle-song/bottle_song.go
package bottlesong

import (
	"fmt"
	"strings"
)

func Recite(startBottles, takeDown int) []string {
	var result []string

	for i := 0; i < takeDown; i++ {
		current := startBottles - i
		verse := generateVerse(current)
		result = append(result, verse...)
		if i != takeDown-1 {
			result = append(result, "")
		}
	}

	return result
}

func generateVerse(n int) []string {
	numWord := numberToWord(n)
	bottles := "bottles"
	if n == 1 {
		bottles = "bottle"
	}

	nextNum := n - 1
	nextNumWord := numberToWord(nextNum)
	nextBottles := "bottles"
	if nextNum == 1 {
		nextBottles = "bottle"
	}

	line1 := fmt.Sprintf("%s green %s hanging on the wall,", numWord, bottles)
	line2 := line1
	line3 := "And if one green bottle should accidentally fall,"
	line4 := fmt.Sprintf("There'll be %s green %s hanging on the wall.", strings.ToLower(nextNumWord), nextBottles)

	return []string{line1, line2, line3, line4}
}

func numberToWord(n int) string {
	switch n {
	case 10:
		return "Ten"
	case 9:
		return "Nine"
	case 8:
		return "Eight"
	case 7:
		return "Seven"
	case 6:
		return "Six"
	case 5:
		return "Five"
	case 4:
		return "Four"
	case 3:
		return "Three"
	case 2:
		return "Two"
	case 1:
		return "One"
	case 0:
		return "No"
	default:
		return ""
	}
}

File: src/2025-02-24-deepclaude-rust/java/exercises/practice/bottle-song/src/main/java/BottleSong.java
class BottleSong {

    String recite(int startBottles, int takeDown) {
        StringBuilder verses = new StringBuilder();
        for (int i = 0; i < takeDown && startBottles - i > 0; i++) {
            int current = startBottles - i;
            String verse = generateVerse(current);
            if (verses.length() > 0) {
                verses.append("\n\n");
            }
            verses.append(verse);
        }
        return verses.toString();
    }

    private String generateVerse(int current) {
        String currentLine = getBottleLine(current);
        int next = current - 1;
        String nextLine = getBottleLine(next);
        
        String line1 = currentLine + " hanging on the wall,";
        String line2 = currentLine + " hanging on the wall,";
        String line3 = "And if one green bottle should accidentally fall,";
        String line4 = "There'll be " + nextLine.toLowerCase() + " hanging on the wall.";
        
        return String.join("\n", line1, line2, line3, line4);
    }

    private String getBottleLine(int n) {
        if (n == 0) {
            return "No green bottles";
        }
        String numberWord;
        switch (n) {
            case 10: numberWord = "Ten"; break;
            case 9: numberWord = "Nine"; break;
            case 8: numberWord = "Eight"; break;
            case 7: numberWord = "Seven"; break;
            case 6: numberWord = "Six"; break;
            case 5: numberWord = "Five"; break;
            case 4: numberWord = "Four"; break;
            case 3: numberWord = "Three"; break;
            case 2: numberWord = "Two"; break;
            case 1: numberWord = "One"; break;
            default: return "";
        }
        String bottles = (n == 1) ? "bottle" : "bottles";
        return numberWord + " green " + bottles;
    }

}
